explore/explore_internal.h

```diff
@@ -0,0 +1,347 @@
+  }
+
+  template<typename It>
+  int enforce_minimum_probability(float minimum_uniform, bool update_zero_elements, It pdf_first, It pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    // iterators don't support <= in general
+    if (pdf_first == pdf_last || pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
+
+	  size_t num_actions = pdf_last - pdf_first;
+
+    if (minimum_uniform > 0.999) // uniform exploration
+    {
+      size_t support_size = num_actions;
+      if (!update_zero_elements)
+      {
+        for (It d = pdf_first; d != pdf_last; ++d)
+          if (*d == 0)
+            support_size--;
+      }
+
+	    for (It d = pdf_first; d != pdf_last; ++d)
+        if (update_zero_elements || *d > 0)
+          *d = 1.f / support_size;
+
+      return S_EXPLORATION_OK;
+    }
+
+    minimum_uniform /= num_actions;
+    float touched_mass = 0.;
+    float untouched_mass = 0.;
+    uint16_t num_actions_touched = 0;
+
+	  for (It d = pdf_first; d != pdf_last; ++d)
+    {
+      auto& prob = *d;
+      if ((prob > 0 || (prob == 0 && update_zero_elements)) && prob <= minimum_uniform)
+      {
+        touched_mass += minimum_uniform;
+        prob = minimum_uniform;
+        ++num_actions_touched;
+      }
+      else
+        untouched_mass += prob;
+    }
+
+    if (touched_mass > 0.)
+    {
+      if (touched_mass > 0.999)
+      {
+        minimum_uniform = (1.f - untouched_mass) / (float)num_actions_touched;
+        for (It d = pdf_first; d != pdf_last; ++d)
+        {
+          auto& prob = *d;
+          if ((prob > 0 || (prob == 0 && update_zero_elements)) && prob <= minimum_uniform)
+            prob = minimum_uniform;
+        }
+      }
+      else
+      {
+        float ratio = (1.f - touched_mass) / untouched_mass;
+        for (It d = pdf_first; d != pdf_last; ++d)
+          if (*d > minimum_uniform)
+            *d *= ratio;
+      }
+    }
+    
+    return S_EXPLORATION_OK;
+  }
+
+  template<typename It>
+  int enforce_minimum_probability(float minimum_uniform, bool update_zero_elements, It pdf_first, It pdf_last)
+  {
+	  typedef typename std::iterator_traits<It>::iterator_category pdf_category;
+
+	  return enforce_minimum_probability(minimum_uniform, update_zero_elements, pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename It>
+  int sample_after_normalizing(uint64_t seed, It pdf_first, It pdf_last, uint32_t& chosen_index, std::input_iterator_tag pdf_category)
+  {
+    if (pdf_first == pdf_last || pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
+    // Create a discrete_distribution based on the returned weights. This class handles the
+    // case where the sum of the weights is < or > 1, by normalizing agains the sum.
+    float total = 0.f;
+    for (It pdf = pdf_first; pdf != pdf_last; ++pdf)
+    {
+      if (*pdf < 0)
+        *pdf = 0;
+
+      total += *pdf;
+    }
+
+    // assume the first is the best
+    if (total == 0)
+    {
+      chosen_index = 0;
```

Comment:
In this case, the returned distribution is wrong.   Make *pdf_first = 1?

Reply:
done